bes: add retry attempt to build tool stream request

sluongng opened the pull request:
Today, all build events sent from Bazel should be delivered reliably
and ack-ed by BES service, otherwise, Bazel will attempt to re-uploads
the missing events.

Currently, bazel and BES service communicate using a bi-directional
GRPC stream, which is subjected to network failures. Especially in the
case where (1) BES service received an event and sent ACK to Bazel, but
(2) the ACK message is lost in transit, subsequently causing (3) Bazel
to retry sending the same event, it would be hard for BES service to
tell why an event is being sent out of the expected sequence (i.e. not
receiving BuildStarted event).

To help BES service differentiate between original events and events
sent because of retry attempts, add a new `retry_attempt` field to
PublishBuildToolEventStreamRequest in publish_build_event.proto.
A value of zero for `retry_attempt` should signify a fresh new event
while a value greater than zero should signify an event being sent
in a retry attempt.

sluongng commented:
cc: @michaeledgar

bduffany commented on third_party/googleapis/google/devtools/build/v1/publish_build_event.proto:
```diff
@@ -184,4 +184,8 @@ message PublishBuildToolEventStreamRequest {
   // i.e. the first event in the stream.
   bool check_preceding_lifecycle_events_present = 7;
+
+  // The number of retry attempts has been made to publish the unique event.
+  // Starts from 0.
+  int32 retry_attempt = 8;
```
I think this should be `attempt_number` and start from 1 so that the server can distinguish between "this is the first attempt" and "this is an old version of bazel that doesn't support this field", which are currently both represented as 0.

michaeledgar reviewed (changes requested):
This is a helpful proposal! I think we should clarify the semantics a bit before submission.

sluongng committed:
bes: add retry attempt to build tool stream request

Today, all build events sent from Bazel should be delivered reliably
and ack-ed by BES service, otherwise, Bazel will attempt to re-uploads
the missing events.

Currently, bazel and BES service communicate using a bi-directional
GRPC stream, which is subjected to network failures. Especially in the
case where (1) BES service received an event and sent ACK to Bazel, but
(2) the ACK message is lost in transit, subsequently causing (3) Bazel
to retry sending the same event, it would be hard for BES service to
tell why an event is being sent out of the expected sequence (i.e. not
receiving BuildStarted event).

To help BES server differentiate between original events and events
sent because of retry attempts, add a new retry_attempt_number field
to PublishBuildToolEventStreamRequest in publish_build_event.proto.
A value 1 for retry_attempt_number should signify a fresh new event
while a value greater than 1 should signify an event being sent in a
retry attempt. The default value 0 means that the client does not
support this field.

sluongng commented:
@michaeledgar Sorry for the slow turnaround time. 🙏 

I have fixed up the PR according to your feedback + rebased it to the latest master.

The tests should all pass now except for some infra failures on BazelCI side, which we could retry later.

Could you please take another look at this?

meisterT commented:
@michaeledgar can you please have another look at this?
